pkg/kubectl/cmd/util/factory.go

```diff
@@ -597,21 +613,22 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 	}
 }
 
-// GetFirstPod returns the first pod of an object from its namespace and selector and the number of matching pods
+// GetFirstPod returns a pod matching the namespace and label selector
+// and the number of all pods that match the label selector.
 func GetFirstPod(client *client.Client, namespace string, selector labels.Selector) (*api.Pod, int, error) {
+	options := api.ListOptions{LabelSelector: selector}
 	var pods *api.PodList
-	for pods == nil || len(pods.Items) == 0 {
+	// TODO: Make timeout configurable
+	if err := wait.PollImmediate(2*time.Second, 20*time.Second, func() (bool, error) {
```

Comment:
I'd much rather you pick up @smarterclayton's `Until` implementation that is watch based from origin.